genie-web/src/main/java/com/netflix/genie/web/agent/apis/rpc/v4/endpoints/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -142,333 +151,630 @@ public GRpcAgentFileStreamServiceImpl(
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public Optional<DirectoryManifest> getManifest(final String jobId) {
-        final ControlStreamObserver streamObserver = this.jobIdControlStreamMap.get(jobId);
-        if (streamObserver != null) {
-            return Optional.ofNullable(streamObserver.manifestRef.get());
-        }
-        log.warn("Stream Record not found for job id: {}", jobId);
-        return Optional.empty();
+        return Optional.ofNullable(this.controlStreamsManager.getManifest(jobId));
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public StreamObserver<AgentManifestMessage> sync(final StreamObserver<ServerControlMessage> responseObserver) {
-        log.info("New agent control stream established");
-        return new ControlStreamObserver(this, responseObserver);
+        return this.controlStreamsManager.handleNewControlStream(responseObserver);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public StreamObserver<AgentFileMessage> transmit(final StreamObserver<ServerAckMessage> responseObserver) {
-        log.info("New file transfer stream established");
-        final FileTransferStreamObserver fileTransferStreamObserver =
-            new FileTransferStreamObserver(this, responseObserver);
-
-        this.pendingTransferObserversSet.add(fileTransferStreamObserver);
-
-        // Schedule a timeout for this transfer to start (first chunk received)
-        this.taskScheduler.schedule(
-            () -> {
-                final boolean removed = pendingTransferObserversSet.remove(fileTransferStreamObserver);
-                if (removed) {
-                    fileTransferStreamObserver.responseObserver.onError(
-                        new GenieTimeoutException("Timeout waiting for transfer to begin")
-                    );
-                }
-            },
-            Instant.now().plusMillis(FILE_TRANSFER_BEGIN_TIMEOUT_MILLIS)
-        );
-
-        return fileTransferStreamObserver;
-    }
-
-    private void registerControlStream(final String jobId, final ControlStreamObserver controlStreamObserver) {
-        final ControlStreamObserver previousObserver = this.jobIdControlStreamMap.put(jobId, controlStreamObserver);
-        if (previousObserver != null) {
-            // In theory, this cannot happen
-            log.warn("Found an previous observer for job id: {}", jobId);
-        }
+        return this.transferManager.handleNewTransferStream(responseObserver);
     }
 
-    private void unregisterControlStream(final String jobId, final ControlStreamObserver controlStreamObserver) {
-        final boolean removed = this.jobIdControlStreamMap.remove(jobId, controlStreamObserver);
-        if (!removed) {
-            log.warn("Could not remove observer for job: {}, not found in map", jobId);
-        }
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void contribute(final Info.Builder builder) {
```

Comment:
👍

Reply:
I had to pull this out in the next commit (leaving it in a separate commit because i hope to restore it at some point).

Something with having this here creates a circular dependency in Spring initialization and the app doesn't start.